clients/spark/src/main/scala/io/treeverse/clients/S3ClientBuilder.scala

```diff
@@ -62,8 +69,22 @@ object S3ClientBuilder extends S3ClientBuilder {
     //     Possibly pre-generate a FileSystem to access the desired bucket,
     //     and query for its credentials provider.  And cache them, in case
     //     some objects live in different buckets.
-    val credentialsProvider =
-      if (hc.get(Constants.AWS_CREDENTIALS_PROVIDER) == AssumedRoleCredentialProvider.NAME) {
+    val credentialsProvider: Option[AWSCredentialsProvider] = {
+      val roleArn = hc.get(Constants.ASSUMED_ROLE_ARN, null)
+
+      if (roleArn != null && !roleArn.isEmpty) {
+        try {
+          val sessionName = s"lakefs-gc-${UUID.randomUUID().toString}"
+          Some(
+            new STSAssumeRoleSessionCredentialsProvider.Builder(roleArn, sessionName)
+              .withLongLivedCredentialsProvider(new DefaultAWSCredentialsProviderChain())
+              .build()
+          )
+        } catch {
+          case _: Exception =>
+            None
+        }
+      } else if (hc.get(Constants.AWS_CREDENTIALS_PROVIDER) == AssumedRoleCredentialProvider.NAME) {
```

Comment:
simplify this code - we first initialize credentials provider (static or role based)
later setup sts using the configured credentials provider as base.

what happen today in case no provider is configured? does the code uses the DefaultASWSCredentialsProviderChain? if this is the case we can skip setting the `withLngLivedCredetntialsProvider`. if we do no return DefaultAWSCredentialsProviderChain as fallback we should not use it for sts if possible.

Reply:
My comment was related to the order of the initialization. We first setup provider, which can be 'none', after that we setup STS based on role. Setting up the STS uses the base provider.
1. The current code setup default credentials provider in case of none - is it by design? why we do not return the default so far and fallback to None (without STS)?
2. In case we need to setup STS the code should build the session using the base provider.
3. In case of any error we do not log the exception and just return none.